library/common/http/dispatcher.cc

```diff
@@ -478,19 +478,38 @@ const LowerCaseString ClusterHeader{"x-envoy-mobile-cluster"};
 const std::string BaseCluster = "base";
 const std::string BaseWlanCluster = "base_wlan";
 const std::string BaseWwanCluster = "base_wwan";
+const LowerCaseString H2UpstreamHeader{"x-envoy-mobile-upstream-protocol"};
+const std::string H2Suffix = "_h2";
+const std::string BaseClusterH2 = BaseCluster + H2Suffix;
+const std::string BaseWlanClusterH2 = BaseWlanCluster + H2Suffix;
+const std::string BaseWwanClusterH2 = BaseWwanCluster + H2Suffix;
 } // namespace
 
-void Dispatcher::attachPreferredNetwork(HeaderMap& headers) {
+void Dispatcher::setDestinationCluster(HeaderMap& headers) {
+
+  // Determine upstream protocol. Use http2 if selected for explicitly, otherwise (any other value,
+  // absence of value) select http1.
+  // TODO(junr03): once http3 is available this would probably benefit from some sort of struct that
+  // maps to appropriate cluster names. However, with only two options (http1/2) this suffices.
+  bool use_h2_cluster{};
+  const HeaderEntry* entry = headers.get(H2UpstreamHeader);
+  if (entry) {
+    if (entry->value() == "http2") {
+      use_h2_cluster = true;
```

Comment:
Also seems like this should be `use_h2_cluster = entry->value() == "http2"`

Reply:
I want the else for the assertion. I guess I could do 

```
use_h2_cluster = entry->value() == "http2"
if (!use_h2_cluster) { ASSERT(entry->value()->getStringView() == "http1"); }
```

six of one half a dozen of the other?

Reply:
I think what you have now is fine